### PR TITLE
log warnings when state processor threads are stuck

### PR DESCRIPTION
**Pull request by efonsell**

### CODE REVIEW

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowDispatcher.java [hunk 1]
```diff
@@ -55,6 +55,9 @@ public void run() {
 
           if (!shutdownRequested) {
             executorRecovery.tick();
+            if (stateProcessorFactory.getPotentiallyStuckProcessors() == executor.getThreadCount()) {
+              logger.warn("All state processor threads are potentially stuck.");
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowDispatcher.java]: Once things get stuck, do we need the logging at every polling interval? 

Maybe use this https://github.com/NitorCreations/nflow/blob/master/nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/util/PeriodicLogger.java

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java [hunk 2]
```diff
@@ -38,7 +51,23 @@ public WorkflowStateProcessorFactory(WorkflowDefinitionService workflowDefinitio
...
+        workflowInstancePreProcessor, env, processingInstances, listeners);
+  }
+
+  public int getPotentiallyStuckProcessors() {
+    DateTime now = now();
+    int potentiallyStuck = 0;
+    synchronized (processingInstances) {
+      for (Entry<Integer, DateTime> entry : processingInstances.entrySet()) {
+        int processingTimeMinutes = minutesBetween(entry.getValue(), now).getMinutes();
+        if (processingTimeMinutes > 5) {
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java]: Make value configurable.

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java [hunk 3]
```diff
@@ -23,6 +34,8 @@
   private final Environment env;
   @Autowired(required = false)
   protected WorkflowExecutorListener[] listeners = new WorkflowExecutorListener[0];
+  final Map<Integer, DateTime> processingInstances = synchronizedMap(new HashMap<Integer, DateTime>());
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java]: LinkedHashMap

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java (hunk 2 above)
Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java]: And I think that the unit should be in seconds.

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java (hunk 3 above)
Comment (gmokki) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java]: Why LinkedHashMap? Nothing in the code relies on the insertion order and triple-memory usage of LinkedHashMap with much slower operations is unneeded overhead.
I propose we just use ConcurrentHashMap -> no need to synchronize even in getPotentiallyStuckProcessors()

Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java]: With linkedhashmap order where stuck threads are logged ( https://github.com/NitorCreations/nflow/pull/184/commits/f7b3e030fdb0f72a946386d18ea4a03132f510b5#diff-ec7adb4d50d2123d1c7a446126265b75R65 ) would stay the same. But ConcurrentHashMap would be better.

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowDispatcher.java (hunk 1 above)
Comment (gmokki) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowDispatcher.java]: Good idea. I just added a pull request to add support for logging warnings through PeriodicLogger
https://github.com/NitorCreations/nflow/pull/185

Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowDispatcher.java]: I merged #185 to master. 

Also, would it be interesting to print stuck threads count also when everything is not stuck.
Something like this:

       7 of 10 state processor threads are potentially stuck (processing more than N seconds).

Comment (efonsell) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowDispatcher.java]: Done

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java (hunk 2 above)
Comment (efonsell) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java]: Done

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java (hunk 3 above)
Comment (efonsell) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessorFactory.java]: Done

**Pull request merged**